plutus-chain-index/src/Control/Concurrent/STM/TBMQueue.hs

```diff
@@ -0,0 +1,77 @@
+module Control.Concurrent.STM.TBMQueue
+  ( TBMQueue
+  , newTBMQueue
+  , newTBMQueueIO
+  , writeTBMQueue
+  , isFullTBMQueue
+  , flushTBMQueue
+  , sizeTBMQueue
+  ) where
+
+import Control.Concurrent.STM (STM, retry)
+import Control.Concurrent.STM.TQueue (TQueue, flushTQueue, isEmptyTQueue, newTQueue, newTQueueIO, writeTQueue)
+import Control.Concurrent.STM.TVar (TVar, newTVar, newTVarIO, readTVar, writeTVar)
+import Control.Monad (when)
+import Numeric.Natural (Natural)
+
+
+-- | 'TBMQueue' is an abstract type representing a bounded FIFO channel with a custom measure function.
+-- 'TBMQueue' with a measure function 'const 1' should be equivalent to 'TBQueue'.
+data TBMQueue a = TBMQueue !(TQueue a)
+                           {-# UNPACK #-} !(TVar Natural) -- Current size
```

Comment:
It unpacks the representation of the data — instead of using a pointer it tries to use unboxed data.

https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/exts/pragmas.html#unpack-pragma:

>The UNPACK indicates to the compiler that it should unpack the contents of a constructor field into the constructor itself, removing a level of indirection.